More invalidation fixes

timholy opened the pull request:
These fix a fair number of the Base invalidations in #35922. If all these seem fine it could be merged, but many do arise from partial inference, and if we implement some of the suggestions in that issue then the changes to the following may not be so urgent:

- ~`convert(::Type{T}, x::AbstractDict) where T<:AbstractDict`~
- `cmp(a::AbstractString, b::AbstractString)`

I doubt either of these would cause trouble, but I'm happy to drop those if Jeff or others think they are fixable by other means. To me, the rest seem like stuff we probably want.

timholy committed:
Prevent more ! invalidations from IOContexts

All of these "fields" are documented to be Bool, so this should not
break any code and may improve error messages in case of user error.

Usages like `if limit...`, `limit ? a : b`, or `limit && return nothing`
do not need these fixes, since these are baked into the language
rather than being handled by dispatch.

timholy committed:
Add some typeasserts to reduce invalidation

timholy committed:
Add `show` method for `RawFD`

This prevents invalidation via `convert(::Type{Ptr{Cint}}, fd)`.

timholy committed:
Add a defensive ambiguity-resolving method

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -609,7 +609,7 @@ filtervalid(names) = filter(x->!occursin(r"#", x), map(string, names))
 accessible(mod::Module) =
     [filter!(s -> !Base.isdeprecated(mod, s), names(mod, all = true, imported = true));
      map(names, moduleusings(mod))...;
-     collect(keys(Base.Docs.keywords))] |> unique |> filtervalid
+     collect(keys(Base.Docs.keywords))]::Vector{Symbol} |> unique |> filtervalid
```
```suggestion
    Symbol[filter!(s -> !Base.isdeprecated(mod, s), names(mod, all = true, imported = true));
      map(names, moduleusings(mod))...;
      collect(keys(Base.Docs.keywords))] |> unique |> filtervalid
```

JeffBezanson commented on base/abstractdict.jl:
```diff
@@ -516,7 +516,7 @@ convert(::Type{T}, x::T) where {T<:AbstractDict} = x
 function convert(::Type{T}, x::AbstractDict) where T<:AbstractDict
     h = T(x)
-    if length(h) != length(x)
+    if length(h)::Int != length(x)::Int
```
I'm sure this is true in practice but it doesn't seem strictly correct to assert it. We don't know what `x` is.

JeffBezanson commented on base/arrayshow.jl:
```diff
@@ -371,7 +371,7 @@ preceded by `prefix`, supposed to encode the type of the elements.
 function _show_nonempty(io::IO, X::AbstractMatrix, prefix::String)
     @assert !isempty(X)
-    limit = get(io, :limit, false)::Bool
+    limit::Bool = get(io, :limit, false)
```
I would be surprised if this is really better --- this way it will have to do `convert(Bool, ::Any)` instead of just asserting the type.

timholy replied:
Here I was just being consistent with the surrounding code; someone went through previously and put assertions on all the `:limit` cases, and there were all of the `limit::Bool = ` style. The difference triggered just enough OCD that I changed it.

JeffBezanson replied:
Aha, then all of those should probably be changed the other way! :)

timholy commented:
If folks want to keep these commits separate for easier reversion in case of trouble I can rebase, or feel free to squash.

JeffBezanson commented on base/Enums.jl:
```diff
@@ -29,7 +29,7 @@ Base.print(io::IO, x::Enum) = print(io, Symbol(x))
 function Base.show(io::IO, x::Enum)
     sym = Symbol(x)
-    if !get(io, :compact, false)
+    if !(get(io, :compact, false)::Bool)
```
Fortunately `::` already binds tighter than `!`, but I can see how that might not be obvious to code readers.

timholy replied:
i can change it, but I felt that obviousness was worth a few extra characters.

timholy committed:
Ensure all `eof` methods return Bool

timholy commented:
Seeing if I can trigger tests.

timholy committed:
Fix inference of `Docs.meta(::Module)`

timholy committed:
Prevent collect dispatch on Expr.args

If you know you're building args of any Expr, you might as well start out
collecting the arguments to a Vector{Any}, rather than using the clever
type-narrowing machinery of collect/map/broadcast.

timholy committed:
Hide CoreLogging.catch_exceptions behind runtime dispatch

timholy commented:
Barring objections, I'll merge this sometime early in the work week.
